hibernate-core/src/main/java/org/hibernate/cache/internal/NaturalIdCacheKey.java

```diff
@@ -38,7 +38,7 @@ public NaturalIdCacheKey(Object naturalIdValues, EntityPersister persister, Shar
 	}
 
 	public NaturalIdCacheKey(Object naturalIdValues, EntityPersister persister, String entityName, SharedSessionContractImplementor session) {
-		this( persister.getNaturalIdMapping().disassemble( naturalIdValues, session ), entityName, session.getTenantIdentifier(), persister.getNaturalIdMapping().calculateHashCode( naturalIdValues ) );
+		this( persister.getNaturalIdMapping().disassemble( naturalIdValues, session ), entityName, session.getTenantIdentifier(), persister.getNaturalIdMapping().calculateHashCode( naturalIdValues, session ) );
```

Comment:
You effectively do the disassemble twice, not sure how that is more performant.  Would be better to do that disassembly once.  Maybe make the disassembled state be a pre-condition to `#calculateHashCode`

Reply:
Another (probably better) option is to calculate the hash-code using the attributes

Reply:
Do you mean without disassembling ?
That's what I did initially but it broke a unit test (it would require to implement equals()/hashCode() for every attribute).

Reply:
I presume a valid correction will be brought by #6248 :)